dask_kubernetes/operator/controller/networking.py

```diff
@@ -0,0 +1,176 @@
+    service_name_resolution_retries=20,
+    port_name="tcp-comm",
+    local_port=None,
+):
+    """Take a service object and return the scheduler address."""
+    [port] = [
+        port.port
+        for port in service.spec["ports"]
+        if port["name"] == service.name or port["name"] == port_name
+    ]
+    if not port_forward_cluster_ip:
+        with suppress(socket.gaierror):
+            # Try to resolve the service name. If we are inside the cluster this should succeed.
+            host = f"{service.name}.{service.namespace}"
+            if await _is_service_available(
+                host=host, port=port, retries=service_name_resolution_retries
+            ):
+                return f"tcp://{host}:{port}"
+
+    # If the service name is unresolvable, we are outside the cluster and we need to port forward the service.
+    host = "localhost"
+
+    port = await port_forward_service(service, port, local_port)
+    return f"tcp://{host}:{port}"
+
+
+async def get_external_address_for_scheduler_service(
+    core_api,
+    service,
+    port_forward_cluster_ip=None,
+    service_name_resolution_retries=20,
+    port_name="tcp-comm",
+    local_port=None,
+):
+    """Take a service object and return the scheduler address."""
+    [port] = [
+        port["port"]
+        for port in service.spec["ports"]
+        if port["name"] == service.name or port["name"] == port_name
+    ]
+    if service.spec["type"] == "LoadBalancer":
+        lb = service.status["load_balancer.ingress"][0]
+        host = lb.hostname or lb.ip
+    elif service.spec["type"] == "NodePort":
+        nodes = await core_api.list_node()
+        host = nodes.items[0].status.addresses[0].address
+    elif service.spec["type"] == "ClusterIP":
+        if not port_forward_cluster_ip:
+            with suppress(socket.gaierror):
+                # Try to resolve the service name. If we are inside the cluster this should succeed.
+                host = f"{service.name}.{service.namespace}"
+                if await _is_service_available(
+                    host=host, port=port, retries=service_name_resolution_retries
+                ):
+                    return f"tcp://{host}:{port}"
+
+        # If the service name is unresolvable, we are outside the cluster and we need to port forward the service.
+        host = "localhost"
+
+        port = await port_forward_service(service, port, local_port)
+    return f"tcp://{host}:{port}"
+
+
+async def _is_service_available(host, port, retries=20):
+    for i in range(retries):
+        try:
+            return await asyncio.get_event_loop().getaddrinfo(host, port)
+        except socket.gaierror as e:
+            if i >= retries - 1:
+                raise e
+            await asyncio.sleep(0.5)
+
+
+async def port_forward_service(service, remote_port, local_port=None):
+    async with PortForward(service, remote_port, local_port) as port:
+        print(f"Forwarding to port {port}")
+        return port
```

Comment:
If you do `async with` the port forward will close when the function returns.

```suggestion
    pf = service.portforward(remote_port, local_port)
    port = await pf.start()
    print(f"Forwarding to port {port}")
    return port
```

We might also want to return `pf` and hold a reference to it somewhere but I'm not sure.